### PR TITLE
AAE-36650 Add dependabot-coverage-check action to validate GitHub Action

### PR DESCRIPTION
**Pull request by wojciech-piotrowiak**
…ions in dependabot.yml

### Checklist

- Jira Reference (also in PR title):
- [README](https://github.com/Alfresco/alfresco-build-tools/blob/master/docs/README.md) updated after adding/changing behaviour of an action
- Proposed version increment for [release](https://github.com/Alfresco/alfresco-build-tools/blob/master/docs/README.md#release):
  - [ ] Patch (bugfix)
  - [X] Minor (new feature)
  - [ ] Major (breaking changes)
- External PR link where changes has been tested:

### CODE REVIEW

File: .github/actions/dependabot-coverage-check/action.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        # List all action directories
+        actions=$(find .github/actions -mindepth 1 -maxdepth 1 -type d | sed 's|^./||')
+
+        # Extract listed directories from dependabot.yml
+        listed=$(yq '.updates[] | select(.package-ecosystem == "github-actions") | .directories[]' .github/dependabot.yml | sed 's|^/||' | grep '^.github/actions/')
+
+        # Check for missing actions
+        missing=""
+        for action in $actions; do
+        if ! echo "$listed" | grep -qx "$action"; then
```
[the remark this answers is not in the document]
Comment (atchertchian) [reply .github/actions/dependabot-coverage-check/action.yml:22]: @wojciech-piotrowiak do you have a test run somewhere please?

File: .github/actions/dependabot-coverage-check/action.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+name: dependabot-coverage-check
```
Comment (atchertchian) [inline .github/actions/dependabot-coverage-check/action.yml:1]: It would be nice to call this action in the local test.yml workflow: the templating system is supposed to take care of it already during pre-commit, but it can be seen as a test of this action, wdyt?

File: .github/actions/dependabot-coverage-check/action.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (wojciech-piotrowiak) [reply .github/actions/dependabot-coverage-check/action.yml]: Test added

File: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+# dependabot-coverage-check
+# Author: wojciech.piotrowiak
+# Description: Workaround for https://github.com/dependabot/dependabot-core/issues/6345.
+# This script checks if all GitHub Actions in .github/actions are listed in .github/dependabot.yml.
```
Comment (gionn) [inline .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh:6]: update to keep them consistent

anyway I am not so inclined to duplicate these info here given they are already provided in the action.yml

File: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+# Find repository root (where .git directory exists)
```
Comment (gionn) [inline .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh:8]: please drop all of these LLM inflated comments which just state the obvious

File: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  exit 1
+fi
+
+if [ ! -f ".github/dependabot.yml" ]; then
+  echo "Error: .github/dependabot.yml file not found"
+  exit 1
+fi
+
+# Find all directories that contain action.yml files
+actions=$(find .github/actions -name "action.yml" -type f | xargs -I {} dirname {} | sed 's|^./||')
```
[the remark this answers is not in the document]
Comment (gionn) [reply .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh:30]: ensure there is a test for this edge case, if it fails then do what it suggest :D

File: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+# Extract listed directories from dependabot.yml - support both directory and directories
+listed_single=$(yq '.updates[] | select(.package-ecosystem == "github-actions") | .directory' .github/dependabot.yml | sed 's|^/||' | grep '^\.github/actions/' | sed 's|^\./||')
```
[the remark this answers is not in the document]
Comment (gionn) [reply .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh:33]: add a command -v check at the beginning of the script

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -38,3 +38,7 @@ repos:
         language: system
         entry: ./check_readme.sh
         pass_filenames: false
+      - id: check-dependabot-coverage
+        name: Check if all actions are listed in dependabot.yml
+        language: system
+        entry: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh
```
Comment (gionn) [inline .pre-commit-config.yaml:44]: not sure it really make sense here given that we already have `generate-dependabot` before which will never make this to trigger.

File: .github/actions/dependabot-coverage-check/action.yml (hunk 2 above)
Comment (gionn) [reply .github/actions/dependabot-coverage-check/action.yml:1]: there are bats tests, there is really no need to add it elsewhere

Comment (atchertchian) [reply .github/actions/dependabot-coverage-check/action.yml:1]: I meant to call the action from the .github/workflows/test.yml file (after pre-commit, as we're advertizing the action in this repo, not the pre-commit bash script, no? (I mean it could also be documented as is if that's what we want to do).
That's also why I ws saying this action can added to the list of actions that are being called on the common pre-checks actiion in the other "build-tools" repo: this way we do not need to impact all pre-commit configurations for this extra "common" check, we just need to bump the action version. wdyt?

File: .github/actions/dependabot-coverage-check/action.yml
[the remark this answers is not in the document]
Comment (atchertchian) [reply .github/actions/dependabot-coverage-check/action.yml:1]: As people prefer :)

File: .pre-commit-config.yaml (hunk 7 above)
Comment (atchertchian) [reply .pre-commit-config.yaml:44]: that's what I meant in another comment: the action is called, and if it fails, maybe one of those calls is buggy so it's nice to have this in a test.yml pipeline, no?

File: .github/actions/dependabot-coverage-check/action.yml (hunk 2 above)
Comment (atchertchian) [reply .github/actions/dependabot-coverage-check/action.yml:1]: also note that adding bash scripts to .pre-commit local conf is sometimes an issue depending on the developer OS

Comment (wojciech-piotrowiak) [reply .github/actions/dependabot-coverage-check/action.yml]: Added that action to test.yml

File: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (wojciech-piotrowiak) [reply .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh]: Done

File: .github/actions/dependabot-coverage-check/action.yml
[the remark this answers is not in the document]
Comment (wojciech-piotrowiak) [reply .github/actions/dependabot-coverage-check/action.yml:1]: Action renamed

File: .github/actions/dependabot-coverage-check/action.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+author: wojciech.piotrowiak
+description: Workaround for https://github.com/dependabot/dependabot-core/issues/6345. This action checks if all GitHub Actions in .github/actions are listed in .github/dependabot.yml.
```
[the remark this answers is not in the document]
Comment (wojciech-piotrowiak) [reply .github/actions/dependabot-coverage-check/action.yml]: Done

File: .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh (hunk 4 above)
Comment (wojciech-piotrowiak) [reply .github/actions/dependabot-coverage-check/dependabot-coverage-check.sh]: Done

File: .github/actions/dependabot-missing-actions-check/action.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+name: dependabot-coverage-check
+author: wojciech.piotrowiak
+description: Ensure that all GitHub Actions defined in the current repo have managed dependabot updates. Workaround for https://github.com/dependabot/dependabot-core/issues/6345
+runs:
+  using: composite
+
+  steps:
+    - name: Checkout code
+      uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
```
Comment (atchertchian) [inline .github/actions/dependabot-missing-actions-check/action.yml:9]: I would leave that out because most of the time the code will already be checked out on the expected branch/ref to check, we don't want to have to add a boolean input to disable checkout like we do on some actions doing checkout unnecessarily

```suggestion
```

File: docs/README.md [hunk 10]
```diff
@@ -421,6 +422,11 @@ Configures the git username and email to associate commits with the provided ide
 
 The two vars in the previous snippet are [workflow configuration variables](https://github.blog/changelog/2023-01-10-github-actions-support-for-configuration-variables-in-workflows/) that can be created at organization level and shared across different repositories.
 
+### dependabot-missing-actions-check
+
+This action checks if all GitHub Actions are listed in the `.github/dependabot.yml` file.
+This is a workaround for a [GH Link](https://github.com/dependabot/dependabot-core/issues/6345) problem where global configuration does not work.
+
 ### dbp-charts
```
Comment (atchertchian) [inline docs/README.md:430]: you're missing a change to version.txt to do a minor release

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -18,6 +18,8 @@ jobs:
         with:
           allowlist: |
             Alfresco/alfresco-build-tools/
+      - name: Dependabot missing actions check
+        uses: ./.github/actions/dependabot-missing-actions-check
```
Comment (gionn) [inline .github/workflows/test.yml:22]: I will move it to the test job below to make it clear it's really just yet another additional test for it

File: docs/README.md (hunk 10 above)
Comment (wojciech-piotrowiak) [reply docs/README.md:472]: done

File: .github/workflows/test.yml (hunk 11 above)
Comment (wojciech-piotrowiak) [reply .github/workflows/test.yml]: Good point, fixed
